### PR TITLE
Incomplete Index Component System (ICS)

### PR DESCRIPTION
**Pull request by joeltio**
I have created the basic implementation of indexes and components along with their tests in this PR.

The CMakeLists file has also been split up. Each subdirectory of sim/src will have its own CMakeLists file. Each subdirectory represents a library that is supposed to be independent of any other adjacent directories. Currently I have `ics` which stores the core ics concepts and structs and `util` which stores utility types.

Each subdirectory is interpreted as a library in the main CMakeLists file.

### CODE REVIEW

File: sim/CMakeLists.txt [hunk 1]
```diff
@@ -4,8 +4,19 @@ project(bentobox)
 # Source files
 set(SRC_DIR "${CMAKE_CURRENT_SOURCE_DIR}/src")
 set(LIB_DIR "${CMAKE_CURRENT_SOURCE_DIR}/lib")
-set(SOURCES "${SRC_DIR}/main.cpp")
-set(TESTS "${SRC_DIR}/test.cpp")
+set(ICS_DIR "${SRC_DIR}/ics")
+set(UTIL_DIR "${SRC_DIR}/util")
+
+set_property(GLOBAL PROPERTY hello "hello world")
```
Comment (mrzzy) [inline sim/CMakeLists.txt:10]: Is this property used?

Comment (joeltio) [reply sim/CMakeLists.txt:10]: No, was testing out some CMake stuff, I forgot to remove it

Comment (joeltio) [reply sim/CMakeLists.txt]: I have just pushed a commit to remove it

File: sim/CMakeLists.txt [hunk 2]
```diff
@@ -36,8 +52,11 @@ endif(NOT WIN32)
 
 ## Test Dependencies
 # Google test
+set(GOOGLETEST_VERSION 1.8.1)
```
Comment (mrzzy) [inline sim/CMakeLists.txt:55]: Googletest version is tracked by git subtree, so i don't think listing the version here does anything./

File: sim/CMakeLists.txt [hunk 3]
```diff
@@ -4,8 +4,17 @@ project(bentobox)
...
 set(SRC_DIR "${CMAKE_CURRENT_SOURCE_DIR}/src")
 set(LIB_DIR "${CMAKE_CURRENT_SOURCE_DIR}/lib")
-set(SOURCES "${SRC_DIR}/main.cpp")
-set(TESTS "${SRC_DIR}/test.cpp")
+set(ICS_DIR "${SRC_DIR}/ics")
+set(UTIL_DIR "${SRC_DIR}/util")
+
+set(SOURCES
+    "${SRC_DIR}/main.cpp"
+    "${SRC_DIR}/Manager.h"
```
Comment (mrzzy) [inline sim/CMakeLists.txt:12]: Should `Manager.h` be part of sources?

File: sim/CMakeLists.txt [hunk 4]
```diff
@@ -14,6 +23,13 @@ add_executable(${PROJECT_NAME} ${SOURCES})
 add_executable("${PROJECT_NAME}_test" ${TESTS})
 target_include_directories(${PROJECT_NAME} PRIVATE ${SRC_DIR})
 
+# Link libraries
+foreach (lib IN ITEMS util ics)
+    add_subdirectory("${SRC_DIR}/${lib}")
+    target_link_libraries(${PROJECT_NAME} PRIVATE ${lib})
+    target_link_libraries(${PROJECT_NAME}_test PRIVATE -Wl,--whole-archive ${lib}-test -Wl,--no-whole-archive)
+endforeach()
```
Comment (mrzzy) [inline sim/CMakeLists.txt:31]: I don't our project is at a stage where we should have multiple cmake lists. one major downside is that i can no longer run all the tests at once.

File: sim/src/ics/component.h [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    static_assert(Component<archetypes::Component>);
+
+    struct DefaultComponent {
+        bool isActive = true;
+        bool operator==(const DefaultComponent&) const = default;
+    };
+
+    static_assert(Component<DefaultComponent>);
```
Comment (mrzzy) [inline sim/src/ics/component.h:27]: What is the difference between a default component and a normal component?

File: sim/src/Manager.h [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        } else {
+            // The container does not exist
+            auto container = CompContainer<C>();
+            container.add(c);
+            components->insert(container);
+        }
+    }
+
+    template<ics::Component C, ics::Index I>
+    void addIndex(I i) {
```
Comment (mrzzy) [inline sim/src/Manager.h:36]: Assuming refers to index in ECS. If so it would be nice to stick to Entity naming convention.

File: sim/src/ics/compVec.test.cpp [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+using namespace ics;
+
+struct ParentComp : public DefaultComponent {
+    int height;
+};
+
+struct ChildComp : public ParentComp {
+    int width;
+};
```
Comment (mrzzy) [inline sim/src/ics/compVec.test.cpp:15]: Are we going to support component inheritance on the user API?

File: sim/src/Manager.h [hunk 8]
```diff
@@ -0,0 +1,46 @@
+#ifndef BENTOBOX_MANAGER_H
```
Comment (mrzzy) [inline sim/src/Manager.h:1]: `Manager.h` seems  part of ECS, should it be under `ics/`?

File: sim/src/util/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+add_library(
+    ${LIB_NAME}-test
+    ${SOURCES}
+    ${TESTS}
+)
+
+# Cmake could not determine the language used from the files
+# https://stackoverflow.com/questions/11801186/cmake-unable-to-determine-linker-language-with-c
+set_target_properties(${LIB_NAME} PROPERTIES LINKER_LANGUAGE CXX)
```
Comment (mrzzy) [inline sim/src/util/CMakeLists.txt:18]: Is there anyway we can avoid this CmakeList boilerplate for every  folder we create?

File: sim/CMakeLists.txt (hunk 2 above)
Comment (joeltio) [reply sim/CMakeLists.txt:55]: I added this because every time I run configure, I get a warning:
```
CMake Warning at lib/googletest/googletest/CMakeLists.txt:54 (project):
  VERSION keyword not followed by a value or was followed by a value that
  expanded to nothing.
```

File: sim/src/ics/component.h (hunk 5 above)
Comment (joeltio) [reply sim/src/ics/component.h:27]: The `archetypes::Component` is used only for concepts. When defining the index concept, I can only enforce that the argument follows the component concept if I have some type that represents the minimal concept.

The DefaultComponent is just a reference implementation of a component. Whether or not you choose to use it is up to you. As long as your component fulfils the concept, it will work. Currently, the only use of `DefaultComponent` is in tests, so that I can create components without worrying about fulfilling the concept.

In summary: `archetypes::Component` is used to represent the concept, `DefaultComponent` is used as a reference implementation. As they serve separate purposes, they should be separate.

File: sim/src/Manager.h (hunk 6 above)
Comment (joeltio) [reply sim/src/Manager.h:36]: I assume you mean
> Assuming this refers to **entity** in ECS ...
An index is not the same as an entity. An entity in ECS is a single unsigned int which multiple components are associated with. However, an index is a more general idea. An index is the association of any component to anything. Similar to an index in databases, it stores some key which makes retrieval faster.

For example, I can have an `EntityIndex` which works the same as entities and I can have a `TagIndex` which maps a single tag string to components or entities.

File: sim/src/ics/compVec.test.cpp (hunk 7 above)
Comment (joeltio) [reply sim/src/ics/compVec.test.cpp:15]: No, I made this because I was worried about casting issues which I hope to test later on. I had some bad experiences with the previous engine when this happens:
```
compVec<ChildComp> x = ...;
// Somehow needs to be converted to compVec<ParentComp>
compVec<ParentComp> y = x;
// Vectors will have problems accessing the correct memory locations because ChildComp is larger than ParentComp
y.at(3)
```

File: sim/src/Manager.h (hunk 8 above)
Comment (joeltio) [reply sim/src/Manager.h:1]: It is part of ICS, but I can't put it in `ics/` because it uses `typeMap.h`. So, to show the dependencies clearly and because of CMake, I have put it outside.

I can rename it to `ICSManager`, though I am planning to create another namespace for the ICS implementations

File: sim/src/util/CMakeLists.txt (hunk 9 above)
Comment (joeltio) [reply sim/src/util/CMakeLists.txt:18]: I doubt it. I don't think CMakeLists files can export variables to the parent scope. So, the child scope has to handle all the files defined in `SOURCES` and `TESTS`.

File: sim/src/Manager.h (hunk 8 above)
Comment (mrzzy) [reply sim/src/Manager.h:1]: This seems be a symptom of our build system being misconfigured. Lets put this off for now.

> I can rename it to ICSManager, though I am planning to create another namespace for the ICS implementations

`ICSManager` is much clearer.

File: sim/CMakeLists.txt (hunk 2 above)
Comment (mrzzy) [reply sim/CMakeLists.txt:55]: Oops, looks like i misconfigured this according to [this googletest issue](https://github.com/google/googletest/issues/2950). Should be just  `add_subdirectory("${LIB_DIR}/googletest", EXCLUDE_FROM_ALL)`

File: sim/src/ics/compVec.test.cpp (hunk 7 above)
Comment (mrzzy) [reply sim/src/ics/compVec.test.cpp:15]: Parent and child component implies some of superior relationship. Also its not immediately clear how parent component is used us there is no test that uses it. Perhaps a rename and some comments to describe what u are trying todo?

File: sim/src/ics/component.h (hunk 5 above)
Comment (mrzzy) [reply sim/src/ics/component.h:27]: Do you intend for components to be created inheriting `DefaultComponent`? Is there a use for `DefaultComponent` outside of tests? If its only tests than I propose a move to tests and rename to `TestComponent`

File: sim/src/ics/compVec.test.cpp (hunk 7 above)
Comment (joeltio) [reply sim/src/ics/compVec.test.cpp:15]: I think I will just remove one of them and rename accordingly. I can test the issue I described above, later.

File: sim/src/ics/component.h (hunk 5 above)
Comment (joeltio) [reply sim/src/ics/component.h:27]: I intend that the components that I write in C++ will use DefaultComponent. User-defined components will likely be a class as a lot of user-defined attributes should be hidden and accessed only through methods.

File: sim/CMakeLists.txt (hunk 2 above)
Comment (joeltio) [reply sim/CMakeLists.txt]: Fixed in 3016255

File: sim/src/ics/compVec.test.cpp (hunk 7 above)
Comment (joeltio) [reply sim/src/ics/compVec.test.cpp]: Fixed in 59a3e0a

File: sim/src/Manager.h (hunk 8 above)
Comment (joeltio) [reply sim/src/Manager.h]: Fixed in c44326c and 3ef67ff

**Comment by mrzzy** [comment]
LGTM :+1: . Merge at will
